AbstractSQLCompiler.ts

```diff
@@ -370,10 +378,150 @@ const compileSchema = (abstractSqlModel: AbstractSqlModel, engine: Engines, ifNo
 	}
 }
 
+const generateSplit = <T>(src:Array<T>, dst: Array<T>, matchFn: MatchFn<T>): Split<T> => {
+	let modified: Array<Pair<T>> = []
+	return _.reduce(src, (acc, value) => {
+		const match = matchFn(acc.inserted, value)
+		if (_.isUndefined(match)) {
+			return acc
+		} else {
+			acc.inserted = _.without(acc.inserted, match)
+			acc.deleted = _.without(acc.deleted, value)
+			acc.modified.push( { src: value, dst: match } )
+			return acc
+		} 
+	}
+	, { inserted: dst, deleted: src, modified: modified } )
+}
+//  genDiff expects src and dst arrays formatted as described in
+//  genSplit, and three functions to be called in the case of
+//  insertion, deletion or modification.
+//  The match function describes how to find the corresponding T inside a T[].
+//  The result of the match function will be either the element that should be matched to 
+//  the argument or undefined if no such element is found.
+const generateDiff = <T>(insFn: ToSQLFn<T>, delFn: ToSQLFn<T>, modFn: ToSQLFn<Pair<T>>, matchFn: MatchFn<T>, src: Array<T>, dst: Array<T>) => {
+	const split = generateSplit(src, dst, matchFn)
+
+	const diff = _.map(split.modified, modFn)
+	.concat(_.map(split.deleted, delFn))
+	.concat(_.map(split.inserted, insFn))
+	
+	return _.reject(diff, _.isUndefined)
+}
+
+const diffFields = (src: AbstractSqlField[], dst: AbstractSqlField[], mappings: ResourceMap<string>, engine: Engines, ifNotExists: boolean) => {
+	let ifNotExistsStr: string
+	let ifExistsStr: string
+
+	if (ifNotExists) {
+		ifNotExistsStr = 'IF NOT EXISTS '
+		ifExistsStr = 'IF EXISTS ' 
+	} else {
+		ifNotExistsStr = ''
+		ifExistsStr = ''
+	}
+
+	const matchFn: MatchFn<AbstractSqlField> = (fieldArray, field) => {
+		let match = _.find(fieldArray, { fieldName: field.fieldName })
+		if (_.isUndefined(match)) {
+			if (_.isString(mappings[field.fieldName])) {
+				return _.find(fieldArray, { fieldName: mappings[field.fieldName] } )
+			}
+		} else {
+			return match
+		}
+	}
+
+	const insFn: ToSQLFn<AbstractSqlField> = (field) => {
+		// Columns are inserted as NULL for the time being, this is because we have no sensible way to automatically assign a default value
```

Comment:
Some fields do have a default value, so in that case we should make it a NOT NULL column (assuming it's meant to be NOT NULL)